packages/framework/src/Framework/Services/BuildTaskService.php

```diff
@@ -4,89 +4,116 @@
  */
 class BuildTaskService
 {
-    /**
-     * Information for package developers: This offers a hook for packages to add custom build tasks.
-     * Make sure to add the fully qualified class name to the array and doing so by merging the array, not overwriting it.
-     *
-     * @var array<class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>>
-     */
-    public static array $postBuildTasks = [];
+    /** @var array<string, \Hyde\Framework\Features\BuildTasks\BuildTask> */
+    protected array $buildTasks = [];
 
-    protected ?OutputStyle $output;
+    protected ?OutputStyle $output = null;
 
-    public function __construct(?OutputStyle $output = null)
+    public function __construct()
     {
-        $this->output = $output;
-    }
-
-    public function runPostBuildTasks(): void
-    {
-        foreach ($this->getPostBuildTasks() as $task) {
-            $this->run($task);
-        }
+        $this->registerTasks(Config::getArray('hyde.build_tasks', []));
 
-        $this->runIf(GenerateBuildManifest::class, config('hyde.generate_build_manifest', true));
+        $this->registerTasks($this->findTasksInAppDirectory());
     }
 
     /** @return array<class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>> */
-    public function getPostBuildTasks(): array
+    public function getTasks(): array
     {
-        return array_unique(
-            array_merge(
-                config('hyde.post_build_tasks', []),
-                static::findTasksInAppDirectory(),
-                static::$postBuildTasks
-            )
-        );
+        return array_map(function (BuildTask $task): string {
+            return $task::class;
+        }, array_values($this->buildTasks));
     }
 
-    protected static function findTasksInAppDirectory(): array
+    public function runPreBuildTasks(): void
     {
-        $tasks = [];
-
-        foreach (glob(Hyde::path('app/Actions/*BuildTask.php')) as $file) {
-            $tasks[] = str_replace(
-                [Hyde::path('app'), '.php', '/'],
-                ['App', '', '\\'],
-                (string) $file
-            );
+        foreach ($this->buildTasks as $task) {
+            if ($task instanceof RunsBeforeBuild) {
+                $task->run();
+            }
         }
+    }
 
-        return $tasks;
+    public function runPostBuildTasks(): void
+    {
+        foreach ($this->buildTasks as $task) {
+            if ($task instanceof RunsAfterBuild) {
+                $task->run();
+            }
+        }
     }
 
-    public function run(string $task): static
+    /** @param  class-string<\Hyde\Framework\Features\BuildTasks\BuildTask>  $class */
+    public function registerTask(string $class): void
     {
-        $this->runTask(new $task($this->output));
+        $task = new $class($this->output);
+
+        if (! $task instanceof BuildTask) {
+            throw new InvalidArgumentException("BuildTask [$class] must extend the HydeBuildTask class.");
+        }
 
```

Comment:
Throw an exception if the task does not implement a runner interface as the task will never run otherwise, which could be confusing for the developer. Throwing makes it clear that something is indeed wrong.

Reply:
We could also add a warning if both are implemented as the same handle method is run regardless of this, but I don't think that matters enough. We'll put something in the docs.

Reply:
Should we default to implements RunsAfterBuild?